Resampling

OverLordGoldDragon opened the pull request.

OverLordGoldDragon committed:
Resampling

lostanlen commented on kymatio/scattering1d/core/timefrequency_scattering.py:
```diff
@@ -64,7 +64,7 @@ def timefrequency_scattering(
     if average:
         # zero-pad along frequency, map to Fourier domain
-        total_height = 2 ** sc_freq.J_pad
+        total_height = 2 ** sc_freq.J_pad[-1]  # TODO
```
in light of `compute_padding_fr`, i'm pretty sure that this `# TODO` is now done

OverLordGoldDragon replied:
[Nope](https://github.com/kymatio/kymatio/pull/674#issuecomment-812808596)

lostanlen replied:
ah, i see, i had missed that earlier. i'm fine with either implementation, with some preference towards padding `S1` the same as `Y2` as it will avoid duplicating the same code across backends. i don't think that this is critical for performance, is it?

an important aspect to keep in mind is that it would be good if we could disable frequential averaging on S1. (i.e., average U1 in time only and average U2 both in time and frequency)
In this way, S1 would be frequency-equivariant and S2 would be frequency-invariant

OverLordGoldDragon replied:
If we don't address it, it _will_ error for some input combinations - and frankly it'd not add much code.

On averaging U1, currently this only _adds_ information, as we append unaveraged `S1`. Looks to me like the zeroth-order of `S2`, but per my latest comment unsure what'd replace it if we don't average

lostanlen replied:
fine. well, clearly we don't want unanticipated errors so the problem has to be fixed, no matter the cost in complexity.

> On averaging U1, currently this only adds information, as we append unaveraged S1. Looks to me like the zeroth-order of S2, but per my latest comment unsure what'd replace it if we don't average

I don't understand. What i had in mind was a keyword `average_S1_fr=True`, which if set to `False`, would disable frequential averaging at the first order. Something like
`average_S1_fr==True` ==> `S1 = |x * psi| * phi_t * phi_fr`
`average_S1_fr==False` ==> `S1 = |x * psi| * phi_t`

In essence, the first order of JTFS in the case `average_S1_fr=False` would fall back to the first-order of a "plain" temporal `Scattering1D`.
(feature requested by Stéphane Mallat)

OverLordGoldDragon replied:
S. Mallat keeps an eye on Kymatio? My regards, he's [how I found](https://youtu.be/4eyUReyIPXg) Kymatio and implemented much of `ssqueezepy`.

I'm saying this kwarg would only save compute (by a negligible amount), since we [already append](https://github.com/lostanlen/kymatio/blob/e7b04a158ef09e2efda8b529c2ddb173c1669139/kymatio/scattering1d/core/timefrequency_scattering.py#L62) `|x * psi| * phi_t`. Or do I misunderstand?

lostanlen replied:
> S. Mallat keeps an eye on Kymatio?

yes. Glad that you got to know us via his talks!


> since we already append |x * psi| * phi_t.

oops! i hadn't seen that. it's a bit debatable to include both `S_1` and `S1_fr` though.
In their textbook definition, scattering transforms preserve energy and never duplicate it. So from that logic we should either have `S1` (averaged in time) or `S1_fr` (averaged in both time and frequency). The risk of returning both is that users might view them as different features, whereas really one can be obtained deterministically from the other.
Perhaps a solution is to have an `average_fr=True` argument which controls whether `phi_fr` is applied on _both_ `S1` and `S2`? i guess it's easier to explain to newcomers: `average` controls temporal invariance wheras `average_fr` controls frequential invariance.

At this piont, this is more an API question than an implementation question ... there's a design principle called "YAGNI" or "You Ain't Going to Need It", meaning it's important to not anticipate too many features that no one will use. But at the same time i can definitely see a use case for disabling frequential averaging (at least at the first order)

lostanlen replied:
i'm not saying that we need to implement all of this right away but it's worth to remember these aspects when designing the v1 of JTFS in Kymatio

OverLordGoldDragon replied:
The averaging adds frequential invariance for `S_1` coeffs, but we already have a superior version of that in `S_2`; if it doesn't complement `S_1` any other way and add duplication, I agree it should skip.

So with `average_fr` our `psi_t * psi_f` pairs would be `|psi_t * psi_f| * phi_t`? Doable. But doubt we should tie it to `S1_fr`, as we might want (in fact it seems the main use case) `phi_t * phi_f` without `S1_fr`. But that means two extra kwargs...

Whatever we do, there's significant API advantage to keeping output indexed the same, even if it's mostly empty: any other code that tries to work with JTFS's output will have much extra bookkeeping to do. We'll also need all the kwargs to fetch `meta`. For visuals alone it's extra `if-else`s and indexing corrections.

OverLordGoldDragon replied:
Regarding (actual) zeroth-order, I don't think we should scrap it, as we do lose significant information, especially for multi-channel data (dc bias and very low frequencies). Note this information isn't recovered anywhere else.

lostanlen replied:
right, it's true that zeroth-order has a form of information (namely, frequencies below sr/2**J) that isn't present in any other order. that include DC.
i have an audio processing bias, where the DC is irrelevant to auditory perception; but you might argue otherwise for other signal modalities like EEG.
if we can document it and it doesn't interfere too much with JTFS display routines (which assume TF inputs), then by all means let's keep the zeroth order

lostanlen replied:
@OverLordGoldDragon what's the status of this `TODO`?

lostanlen commented on kymatio/scattering1d/core/timefrequency_scattering.py:
```diff
@@ -139,14 +143,18 @@ def timefrequency_scattering(
+        Y_2_c = B.ifft(Y_2_hat)
+
+        if Y_2_arr is None:
+            Y_2_arr = backend.zeros((2**sc_freq.J_pad[-1], Y_2_c.shape[-1]),
```
why initialize `Y_2_arr` with `None` instead of directly this `backend.zeros` array? that would spare an `if` statement in the for loop. Am i missing something important here?

OverLordGoldDragon replied:
See [4 here](https://github.com/kymatio/kymatio/pull/674#issuecomment-815896118)

lostanlen replied:
ah, i see, you need `Y_2_c.shape[-1]`. ok!

lostanlen replied:
maybe remove the `# TODO` in L135 then?

OverLordGoldDragon replied:
It's reminder to take another look if this can be computed easily, may drop with next commit

lostanlen commented on kymatio/scattering1d/core/timefrequency_scattering.py:
```diff
@@ -281,4 +295,12 @@ def _pad_transpose_fft(coeff_list, total_height, B, fft):
     return out
 
 
+def _transpose_fft(coeff_arr, B, fft):
```
I'm not sure that this needs to be its own function. It's only two lines and we don't call it very often.

OverLordGoldDragon replied:
It'll become three with `sc_freq.J_pad[-1]  # TODO`, then we'll remove it entirely once we have `axis=` arg in `fft`

lostanlen commented on kymatio/scattering1d/filter_bank.py:
```diff
@@ -731,3 +731,40 @@ def scattering_filter_factory(J_support, J_scattering, Q, r_psi=math.sqrt(0.5),
+    j0 = 0
+    j0s = [j0]
+    psi1_f_new = [{} for _ in range(len(psi1_f))]
+    # J_pad is ordered lower to greater, so iterate backward then flip
```
a useful comment 👍🏻

lostanlen commented on kymatio/scattering1d/core/timefrequency_scattering.py:
```diff
@@ -103,14 +103,18 @@ def timefrequency_scattering(
-        # zero-pad along frequency, map to Fourier domain
-        total_height = 2 ** sc_freq.J_pad
-        Y_2_hat = _pad_transpose_fft(Y_2_list, total_height, B, B.fft)
+        # zero-pad along frequency, map to Fourier domain  # TODO comment
```
TODO comment ?

OverLordGoldDragon replied:
More like `# TODO improve comment`, i.e. make clearer

lostanlen replied:
that already seems quite clear to me honestly :)

OverLordGoldDragon commented:
Do we agree that it's *re*sampling, not *sub*sampling?

OverLordGoldDragon commented:
Do we agree that it's *re*sampling, not *sub*sampling?

OverLordGoldDragon commented:
Another thing popped up, we have inconsistent output behavior with `average=False`, and need something like this:

```python
        out_S_1.append({'coef': S_1, 'j': (j1,), 'n': (n1,), 's': ()})
    else:
        out_S_1.append({'coef': [], 'j': (), 'n': (), 's': ()})
```
Though, what to put there? `[]` works?

lostanlen commented:
> Though, what to put there? [] works?

No, i don't think an empty list is a good idea here.
But indexing is an entirely separate question so let's take it elsewhere

OverLordGoldDragon committed:
[wip]

OverLordGoldDragon committed:
relocate `else`

OverLordGoldDragon committed:
Create visuals.py

OverLordGoldDragon committed:
stuff

OverLordGoldDragon committed:
Update timefrequency_scattering.py

OverLordGoldDragon committed:
c l e a n

OverLordGoldDragon committed:
Update timefrequency_scattering.py

OverLordGoldDragon committed:
Update timefrequency_scattering.py

OverLordGoldDragon committed:
Update timefrequency_scattering.py

OverLordGoldDragon committed:
Update timefrequency_scattering.py

OverLordGoldDragon committed:
yes

OverLordGoldDragon committed:
staff

OverLordGoldDragon committed:
Update timefrequency_scattering.py

OverLordGoldDragon committed:
Update timefrequency_scattering.py

OverLordGoldDragon commented:
Hardest part done.

 1. I changed variable naming for `_frequency_scattering` and `_joint_lowpass` to be more human readable, should've done from beginning; keeping them in terms of `k` and `j` is very confusing, and they no longer uniquely correspond to padding or subsampling factors to begin with due to "trimming". 
 2. Yet to optimize for readability, some parts could move to `base_frontend`, and there's some duplication. 
 3. Added tests for alignment and shapes, could use shorter `T` for speed?
 4. Need to change `oversampling_fr` kwarg name and values; suggestions? While (implems) 2 and 3 oversample, only 1 is fully equivalent to an integer `oversample_fr=0`; 2 and 3 behaviors aren't attainiable from tweaking `oversample_fr` alone. (instead, if implemented, an integer `oversample_fr` will work _on top of_ current behavior)
 5. Still need to implement wavelet subsampling (currently using resampling).

OverLordGoldDragon committed:
Merge pull request #1 from OverLordGoldDragon/temp

OverLordGoldDragon committed:
Update test_jtfs.py

OverLordGoldDragon committed:
Update test_jtfs.py

lostanlen commented:
@OverLordGoldDragon ok let me know when you want a code review

lostanlen commented on kymatio/scattering1d/core/timefrequency_scattering.py:
```diff
@@ -2,14 +2,15 @@
         x, pad, unpad, backend, J, psi1, psi2, phi, sc_freq,
         pad_left=0, pad_right=0, ind_start=None, ind_end=None, oversampling=0,
-        max_order=2, average=True, size_scattering=(0, 0, 0), out_type='array'):
+        oversampling_fr='auto', max_order=2, average=True,
```
what's `oversampling_fr="auto"`?

OverLordGoldDragon replied:
so not just rename but what the values should be. currently it's binary, so perhaps a boolean

lostanlen commented on kymatio/scattering1d/core/timefrequency_scattering.py:
```diff
@@ -64,23 +65,54 @@ def timefrequency_scattering(
+            reference_total_subsample_so_far = (subsample_equiv_due_to_pad_min +
+                                                0)
+        else:
+            reference_total_subsample_so_far = 0
```
are you saying that _any_ value of `oversampling_fr` other than `"auto"` would lead to the same output? is this consistent with the way `oversampling` works?

lostanlen replied:
said otherwise, if i set `oversampling_fr=0` or `oversampling_fr=1`, do i get the same output?

OverLordGoldDragon replied:
It's not consistent with `oversampling` (see 4 in my latest comment post)

OverLordGoldDragon replied:
We should rename this. The `+ 0` is there for clarity that a particular value is zero rather than absent, and this variable is shown in `_joint_lowpass` (though I could just explicitly declare it and set to zero; this is the duplication I'm still sorting)

lostanlen commented on kymatio/scattering1d/core/timefrequency_scattering.py:
```diff
@@ -64,23 +65,54 @@ def timefrequency_scattering(
         # RFC: should we put placeholders for j1 and n1 instead of empty tuples?
+    else:
+        # TODO what to do here?
+        out_S_1.append({'coef': [], 'j': (), 'n': (), 's': ()})
```
same as Scattering1D. Make a list of the `U_1_m` for different `j`

OverLordGoldDragon replied:
this is already done earlier (which makes me realize, we should append to a different `S_N`). I don't think there's anything to do here but a placeholder

lostanlen commented on kymatio/scattering1d/core/timefrequency_scattering.py:
```diff
@@ -64,23 +65,54 @@ def timefrequency_scattering(
+        S_1_fr_T_hat = B.subsample_fourier(S_1_fr_T_c, 2**lowpass_subsample_fr)
         S_1_fr_T = B.irfft(S_1_fr_T_hat)
 
+        # TODO dedicated ind_start, ind_end and phi
```
DONE ?

OverLordGoldDragon replied:
7 on todo list

lostanlen commented on kymatio/scattering1d/core/timefrequency_scattering.py:
```diff
@@ -164,18 +209,18 @@ def timefrequency_scattering(
 
-    if out_type == 'array':
-        out_S = B.concatenate([x['coef'] for x in out_S])
+    # if out_type == 'array':  # TODO breaks for first-order coeffs
```
remove?

OverLordGoldDragon replied:
This is 12 on [todo list](https://github.com/kymatio/kymatio/issues/712), as we discussed

lostanlen commented on kymatio/scattering1d/core/timefrequency_scattering.py:
```diff
@@ -139,18 +179,23 @@ def timefrequency_scattering(
     # Transform over frequency + low-pass
     # `* psi_f` part of `X * (phi_t * psi_f)`
-    _frequency_scattering(Y_2_hat, j2, -1, k1_plus_k2, commons,
+    _frequency_scattering(Y_2_hat, j2, -1, pad_fr, k1_plus_k2, commons,
```
`spin_down=False`
brilliant !!!!

lostanlen commented on kymatio/visuals.py:
```diff
@@ -0,0 +1,254 @@
+# -*- coding: utf-8 -*-
+"""Convenience visual methods."""
```
i'd rather leave this file outside of the PR for now. but let's keep it for later

OverLordGoldDragon replied:
The plan is to drop it right before merging into `dev`, but keep till then so we can visualize locally

lostanlen commented on kymatio/scattering1d/frontend/base_frontend.py:
```diff
@@ -368,22 +372,128 @@ def _document(cls):
         """This is equivalent to `len(x)` along frequency, which varies across
-        `psi2`; we set this to the longest among them to support all subsamplings
-        and enable concatenation.
+        `psi2`, so we compute for each.
```
very useful function! thanks

lostanlen commented on kymatio/scattering1d/frontend/base_frontend.py:
```diff
@@ -368,22 +372,128 @@ def _document(cls):
+        del self.sc_freq.psi1_f
+
+    def compute_padding_fr(self):
+        """Long story"""  # TODO
```
or else you can rename the method to `_compute_padding_fr` and leave it undocumented ;p

lostanlen replied:
i don't think it'll ever serve in user code

OverLordGoldDragon replied:
To contrary, here we should explain much of the circus in `_frequency_scattering` and `_joint_lowpass`. (or elsewhere, but it must be explained)

lostanlen commented on kymatio/scattering1d/filter_bank.py:
```diff
@@ -731,3 +731,37 @@ def scattering_filter_factory(J_support, J_scattering, Q, r_psi=math.sqrt(0.5),
+            prev_pad = pad
+            for n1 in range(len(psi1_f)):
+                xi, sigma = psi1_f[n1]['xi'], psi1_f[n1]['sigma']
+                # psi1_f_new[n1][j0] = periodize_filter_fourier(psi1_f[n1][0], 2**j0)
```
remove this line?

OverLordGoldDragon committed:
yes

OverLordGoldDragon committed:
Merge pull request #2 from OverLordGoldDragon/temp2

yes

OverLordGoldDragon commented:
1. Rebased `sc_freq`, dedicated own class; was getting too bookkeepy and verbose to distinguish `self` and `self.sc_freq`, and it's large enough to be self-contained
 2. `scattering_filter_factory_fr` now fully computes all filters
 3. misc cleanups

lostanlen commented on kymatio/scattering1d/frontend/base_frontend.py:
```diff
@@ -373,41 +372,128 @@ def _document(cls):
+    def N_fr(self):
+        return self.sc_freq.N_fr
+
+    @property
```
i don't think that it's necessary to expose `max_order_fr` to the user. this is necessarily =1
(3-rd order JTFS is out of scope for now)

lostanlen commented on kymatio/scattering1d/frontend/base_frontend.py:
```diff
@@ -373,41 +372,128 @@ def _document(cls):
+        cls.__doc__ = """Joint time-frequency scattering"""
 
+
+class _FrequencyScatteringBase(ScatteringBase):
```
are you sure that there's 0% of the Scattering1DBase code that could be inherited for this?
It's embarassing to see all these constants sigma0, alpha, Pmax, eps being copy-pasted

lostanlen replied:
otherwise no big deal

OverLordGoldDragon committed:
Faster `compute_minimum_support_to_pad`

OverLordGoldDragon committed:
remove `compute_spin_down_filters`

lostanlen commented:
@OverLordGoldDragon merge?

OverLordGoldDragon commented:
@lostanlen Probably, but will first figure the freq invar thing to ensure it's nothing to do with this PR

OverLordGoldDragon committed:
unpad fix; non-CQT case

OverLordGoldDragon committed:
add padtype

OverLordGoldDragon committed:
`padtype='zero'` JTFS default

OverLordGoldDragon committed:
oversampling_fr -> aligned

OverLordGoldDragon committed:
Add FDTS test

OverLordGoldDragon commented:
Added mainly as discussed; the rest I'm confident on
